File: skills/code-optimizer/references/memory/node/test_code.ts
import { summarizeBad, summarizeGood, type Row } from "./code.ts";

type MemoryUsage = {
  heapUsed: number;
  rss: number;
};

declare const process: {
  memoryUsage(): MemoryUsage;
};

type Sample = {
  elapsedMs: number;
  heapDeltaBytes: number;
  rssBytes: number;
  result: number;
};

function buildRows(count: number, width: number): Row[] {
  const rows: Row[] = [];
  const payload = Array.from(
    { length: width },
    (_, index) => `value-${index}-token`,
  ).join(",");
  for (let index = 0; index < count; index += 1) {
    rows.push({ id: index, payload });
  }
  return rows;
}

function collect(): void {
  const runtime = globalThis as { gc?: () => void };
  runtime.gc?.();
}

function measure(run: () => number): Sample {
  collect();
  const startHeap = process.memoryUsage().heapUsed;
  const start = performance.now();
  const result = run();
  const elapsedMs = performance.now() - start;
  const endMemory = process.memoryUsage();
  return {
    elapsedMs,
    heapDeltaBytes: endMemory.heapUsed - startHeap,
    rssBytes: endMemory.rss,
    result,
  };
}

const rows = buildRows(4000, 25);
const bad = measure(() => summarizeBad(rows));
const good = measure(() => summarizeGood(rows));

if (bad.result !== good.result) {
  throw new Error(`Result mismatch: ${bad.result} !== ${good.result}`);
}

console.table([
  {
    label: "summarizeBad",
    elapsedMs: bad.elapsedMs,
    heapDeltaBytes: bad.heapDeltaBytes,
    rssBytes: bad.rssBytes,
  },
  {
    label: "summarizeGood",
    elapsedMs: good.elapsedMs,
    heapDeltaBytes: good.heapDeltaBytes,
    rssBytes: good.rssBytes,
  },
]);

File: skills/code-optimizer/references/memory/python/test_code.py
from collections.abc import Callable
import gc
import time
import tracemalloc

from code import Row, summarize_bad, summarize_good


def build_rows(count: int, width: int) -> list[Row]:
  payload = ",".join(f"value-{index}-token" for index in range(width))
  return [Row(id=index, payload=payload) for index in range(count)]


def measure(run: Callable[[], int]) -> dict[str, float]:
  gc.collect()
  tracemalloc.start()
  start = time.perf_counter()
  result = run()
  elapsed_ms = (time.perf_counter() - start) * 1000
  current, peak = tracemalloc.get_traced_memory()
  tracemalloc.stop()
  return {
    "result": float(result),
    "elapsed_ms": elapsed_ms,
    "current_bytes": float(current),
    "peak_bytes": float(peak),
  }


rows = build_rows(4000, 25)
bad = measure(lambda: summarize_bad(rows))
good = measure(lambda: summarize_good(rows))

if bad["result"] != good["result"]:
  raise RuntimeError("Memory example results do not match")

for label, sample in (("summarize_bad", bad), ("summarize_good", good)):
  print(
    label,
    {
      "elapsed_ms": sample["elapsed_ms"],
      "current_bytes": sample["current_bytes"],
      "peak_bytes": sample["peak_bytes"],
    },
  )
